In [154]:
import pandas as pd
import plotly.express as px
import re
pd.set_option('display.max_rows', 5000)
pd.set_option('display.max_columns', 5000)
pd.options.display.max_colwidth = 2000

In [155]:
# Functions

def column_name_cleaning(table):
    table.columns = table.columns.str.lower()
    table.columns = [re.sub('(,)|(:)|(-)', '', x) for x in table.columns]
    table.columns = [re.sub('(_)', ' ', x) for x in table.columns]
    table.columns = [x.strip(' ') for x in table.columns]
    
    # patient id column name normalization
    if 'subject number' in table.columns: 
        table = table.rename(mapper={'subject number': 'patient id'}, axis=1)
    elif 'patient id' in table.columns: 
        table = table.rename(mapper={'patient id': 'patient id'}, axis=1)

def str_column_cleaning(table):
    for col in table.columns:
        if table[col].dtype == object:
            table[col] = table[col].str.lower()
            table[col] = table[col].str.replace('(,)|(:)|(-)', '', regex=True)
            table[col] = table[col].str.replace('(_)', ' ', regex=True)
            table[col] = table[col].str.strip(' ')

# Loading

In [156]:
dir = '../data/cerebral_perfusion_data/data_description/'
filename_dat = 'GE-75_data_summary_table.csv'
filename_desc = 'GE-75_data_dictionary.csv'
df = pd.read_csv(dir+filename_dat)
df_desc = pd.read_csv(dir+filename_desc, usecols=range(5))

In [157]:
df.head() #encoding problem, pandas and linux = unknown-8bit, vscode = utf8. Saved it as utf8 on vscode

patient ID study    Group  template  Diabetes Duration HTN or not  HTN2  \
0      S0250  GE75       DM       NaN                7.0        NaN   NaN   
1      S0254  GE75  CONTROL  Modified                0.0        ntn   NaN   
2      S0255  GE75       DM       NaN                3.0        NaN   NaN   
3      S0256  GE75       DM  Modified               23.0        ntn   NaN   
4      s0257  GE75  CONTROL       NaN                0.0        NaN   NaN   

   age     sex  Gender  HEIGHT (M)  MASS (KG)        BMI  global GM vol  \
0   50    Male     NaN       1.790     114.65  35.782279            NaN   
1   69  Female     NaN       1.531      56.90  24.275148        524.915   
2   78  Female     NaN       1.520      65.77  28.466932            NaN   
3   65  Female     NaN       1.575      73.50  29.629630        620.305   
4   55    Male     NaN       1.700      70.31  24.328720            NaN   

   global GM vol covered in template  L superior frontal gyrus (#21) GM  \
0                                NaN                                NaN   
1                            524.736                            22.5098   
2                                NaN                                NaN   
3                            620.189                            28.1071   
4                                NaN                                NaN   

   R superior frontal gyrus (#22) GM  L middle frontal gyrus (#23) GM  \
0                                NaN                              NaN   
1                            23.2694                          18.7579   
2                                NaN                              NaN   
3                            27.5872                          21.8898   
4                                NaN                              NaN   

   R middle frontal gyrus (#24) GM  L inferior frontal gyrus (#25) GM  \
0                              NaN                                NaN   
1                          18.6212                            10.5293   
2                              NaN                                NaN   
3                          22.4723                            11.4893   
4                              NaN                                NaN   

   R inferior frontal gyrus (#26) GM  L precentral gyrus (#27) GM  \
0                                NaN                          NaN   
1                            11.1426                       9.1145   
2                                NaN                          NaN   
3                            11.8439                      10.2783   
4                                NaN                          NaN   

   R precentral gyrus (#28) GM  L middle orbitofrontal gyrus (#29) GM  \
0                          NaN                                    NaN   
1                       8.8643                                 4.2891   
2                          NaN                                    NaN   
3                      11.5083                                 5.6318   
4                          NaN                                    NaN   

   R middle orbitofrontal gyrus (#30) GM  \
0                                    NaN   
1                                 4.4103   
2                                    NaN   
3                                 5.6075   
4                                    NaN   

   L lateral orbitofrontal gyrus (#31) GM  \
0                                     NaN   
1                                  2.6992   
2                                     NaN   
3                                  2.8272   
4                                     NaN   

   R lateral orbitofrontal gyrus (#32) GM  L gyrus rectus (#33) GM  \
0                                     NaN                      NaN   
1                                  2.3685                   1.4723   
2                                     NaN                      NaN   
3                                  2.5678                   2.1167   
4                  

In [158]:
df_desc.head()

SUBJECT NUMBER      S0328      S0342  \
0                          NaN        NaN        NaN   
1                        STUDY      GE-75      GE-75   
2      COMPLETED VISITS STATUS  COMPLETED  COMPLETED   
3  HTN PATIENT MEDICAL HISTORY         NO         NO   
4                      GROUP 2    CONTROL         DM   

           Variable description units  
0                           NaN   NaN  
1                  study number   NaN  
2        completed study status   NaN  
3  hypertension patient history   NaN  
4      group control - diabetes   NaN

# Exploration

The exploration part aims 
- to assess the relevance of the dataset for our use case: explore links between diabetes and cardiovascular diseases in order to the development of a future application to follow  diabetic patients.
- to analyze the consistency with the documentation available for the dataset
- to detect problems in the data: missing value, wrong format, renaming, etc.

3 categories of features have been identified to be potentially interesting: 
- demographics: age, gender, bmi, etc
- lifestyle: physical activity, smoking habits, diet
- medical record: previous cancer, stroke, 

In [159]:
print(df.shape)
print(len(pd.unique(df["patient ID"])))

(88, 1111)
88


In [160]:
print(df_desc.shape)
print(len(pd.unique(df_desc["SUBJECT NUMBER"])))

(1149, 5)
1143


This dataset contains 88 unique patients and 1111 features. This is inconsistent with number of patients found in their paper: 71 TD2M + 76 ctrl 

The description file contains more features than the data file

Let's have an overview of the features:

In [161]:
list(df.columns)

['patient ID',
 'study',
 'Group',
 'template',
 'Diabetes Duration',
 'HTN or not',
 'HTN2',
 'age',
 'sex',
 'Gender',
 'HEIGHT (M)',
 'MASS (KG)',
 'BMI',
 'global GM vol',
 'global GM vol covered in template',
 'L superior frontal gyrus (#21) GM',
 'R superior frontal gyrus (#22) GM',
 'L middle frontal gyrus (#23) GM',
 'R middle frontal gyrus (#24) GM',
 'L inferior frontal gyrus (#25) GM',
 'R inferior frontal gyrus (#26) GM',
 'L precentral gyrus (#27) GM',
 'R precentral gyrus (#28) GM',
 'L middle orbitofrontal gyrus (#29) GM',
 'R middle orbitofrontal gyrus (#30) GM',
 'L lateral orbitofrontal gyrus (#31) GM',
 'R lateral orbitofrontal gyrus (#32) GM',
 'L gyrus rectus (#33) GM',
 'R gyrus rectus (#34) GM',
 'L postcentral gyrus (#41) GM',
 'R postcentral gyrus (#42) GM',
 'L superior parietal gyrus (#43) GM',
 'R superior parietal gyrus (#44) GM',
 'L supramarginal gyrus (#45) GM',
 'R supramarginal gyurs (#46) GM',
 'L angular gyrus (#47) GM',
 'R angular gyrus (#48) GM',


- Features have various names mixing underscores, capslock, spaces, coma, etc.
- Multiple features related to MRI features that are not relevant here: can be filtered with regex

Let's homogenize columns names and rows and then filter MRI features with regex to have a better overview. Let's apply columns names homogenization also to the description file so we can still match description and variables

In [162]:
column_name_cleaning(df)
str_column_cleaning(df)
df.head()

patient id study    group  template  diabetes duration htn or not  htn2  \
0      s0250  ge75       dm       NaN                7.0        NaN   NaN   
1      s0254  ge75  control  modified                0.0        ntn   NaN   
2      s0255  ge75       dm       NaN                3.0        NaN   NaN   
3      s0256  ge75       dm  modified               23.0        ntn   NaN   
4      s0257  ge75  control       NaN                0.0        NaN   NaN   

   age     sex  gender  height (m)  mass (kg)        bmi  global gm vol  \
0   50    male     NaN       1.790     114.65  35.782279            NaN   
1   69  female     NaN       1.531      56.90  24.275148        524.915   
2   78  female     NaN       1.520      65.77  28.466932            NaN   
3   65  female     NaN       1.575      73.50  29.629630        620.305   
4   55    male     NaN       1.700      70.31  24.328720            NaN   

   global gm vol covered in template  l superior frontal gyrus (#21) gm  \
0                                NaN                                NaN   
1                            524.736                            22.5098   
2                                NaN                                NaN   
3                            620.189                            28.1071   
4                                NaN                                NaN   

   r superior frontal gyrus (#22) gm  l middle frontal gyrus (#23) gm  \
0                                NaN                              NaN   
1                            23.2694                          18.7579   
2                                NaN                              NaN   
3                            27.5872                          21.8898   
4                                NaN                              NaN   

   r middle frontal gyrus (#24) gm  l inferior frontal gyrus (#25) gm  \
0                              NaN                                NaN   
1                          18.6212                            10.5293   
2                              NaN                                NaN   
3                          22.4723                            11.4893   
4                              NaN                                NaN   

   r inferior frontal gyrus (#26) gm  l precentral gyrus (#27) gm  \
0                                NaN                          NaN   
1                            11.1426                       9.1145   
2                                NaN                          NaN   
3                            11.8439                      10.2783   
4                                NaN                          NaN   

   r precentral gyrus (#28) gm  l middle orbitofrontal gyrus (#29) gm  \
0                          NaN                                    NaN   
1                       8.8643                                 4.2891   
2                          NaN                                    NaN   
3                      11.5083                                 5.6318   
4                          NaN                                    NaN   

   r middle orbitofrontal gyrus (#30) gm  \
0                                    NaN   
1                                 4.4103   
2                                    NaN   
3                                 5.6075   
4                                    NaN   

   l lateral orbitofrontal gyrus (#31) gm  \
0                                     NaN   
1                                  2.6992   
2                                     NaN   
3                                  2.8272   
4                                     NaN   

   r lateral orbitofrontal gyrus (#32) gm  l gyrus rectus (#33) gm  \
0                                     NaN                      NaN   
1                                  2.3685                   1.4723   
2                                     NaN                      NaN   
3                                  2.5678                   2.1167   
4                  

In [163]:
column_name_cleaning(df_desc)
str_column_cleaning(df_desc)
df_desc.head()

subject number      s0328      s0342  \
0                          NaN        NaN        NaN   
1                        study       ge75       ge75   
2      completed visits status  completed  completed   
3  htn patient medical history         no         no   
4                      group 2    control         dm   

           variable description units  
0                           NaN   NaN  
1                  study number   NaN  
2        completed study status   NaN  
3  hypertension patient history   NaN  
4       group control  diabetes   NaN

In [171]:
brain_measures_re = re.compile('(^r )|(^l )|(#[0-9]+)|(perfusion)')
brain_measures_columns = list(filter(brain_measures_re.search, df.columns))
len(brain_measures_columns)
brain_measures_columns

['l superior frontal gyrus (#21) gm',
 'r superior frontal gyrus (#22) gm',
 'l middle frontal gyrus (#23) gm',
 'r middle frontal gyrus (#24) gm',
 'l inferior frontal gyrus (#25) gm',
 'r inferior frontal gyrus (#26) gm',
 'l precentral gyrus (#27) gm',
 'r precentral gyrus (#28) gm',
 'l middle orbitofrontal gyrus (#29) gm',
 'r middle orbitofrontal gyrus (#30) gm',
 'l lateral orbitofrontal gyrus (#31) gm',
 'r lateral orbitofrontal gyrus (#32) gm',
 'l gyrus rectus (#33) gm',
 'r gyrus rectus (#34) gm',
 'l postcentral gyrus (#41) gm',
 'r postcentral gyrus (#42) gm',
 'l superior parietal gyrus (#43) gm',
 'r superior parietal gyrus (#44) gm',
 'l supramarginal gyrus (#45) gm',
 'r supramarginal gyurs (#46) gm',
 'l angular gyrus (#47) gm',
 'r angular gyrus (#48) gm',
 'l precuneus (#49) gm',
 'r precuneus (#50) gm',
 'l superior occipital gyrus (#61) gm',
 'r superior occipital gyrus (#62) gm',
 'l middle occipital gyrus (#63) gm',
 'r middle occipital gyrus (#64) gm',
 'l infe

In [169]:
print(len(brain_measures_columns))
print(df.shape[1] - len(brain_measures_columns))

915
196


In [170]:
print(len(brain_measures_columns))
print(df.shape[1] - len(brain_measures_columns))

915
196


By filtering those we have now 196 features left

In [172]:
df.drop(brain_measures_columns ,axis='columns', inplace=True)
df.shape #check

(88, 196)

In [174]:
list(df.columns)

['patient id',
 'study',
 'group',
 'template',
 'diabetes duration',
 'htn or not',
 'htn2',
 'age',
 'sex',
 'gender',
 'height (m)',
 'mass (kg)',
 'bmi',
 'global gm vol',
 'global gm vol covered in template',
 'global wm vol',
 'global wm vol covered in template',
 'global csf vol',
 'global csf vol covered in template',
 'global gm covered in template/icv',
 'global wm covered in template/icv',
 'global csf covered in template/icv',
 'global intracranial vol (unit ml)',
 'global gm / icv',
 'global wm / icv',
 'global csf / icv',
 'wmh unregistered',
 'global intracranial vol (cm^3)',
 'wmh registered',
 'wmh registered masked',
 'global vasoreactivity',
 'ethnicity',
 'race',
 'race2',
 'dm nondm stroke',
 'previous tobacco use',
 'current tobacco use',
 'packyears',
 'years',
 'previous alcohol use',
 'current alcohol use',
 'alcohol dose/week',
 'neuropathy autonomic symptoms',
 'dizziness autonomic symptoms',
 'numbness autonomic symptoms',
 'painful feet autonomic symptoms',

In [ ]:
#scores_test = df.loc[:, df.columns.str.contains('(DS)|(^HVLT)|(^VF)|(^Clock)|(^TM)|(Score)')].columns
#pd.unique(df_desc["Variable description"].where(df_desc["SUBJECT NUMBER"].isin(scores_test)))

In [ ]:
filter_1 = '(^R)|(^L)|(Perfusion)|(#[0-9]+)|(^global)|(DS)|(^HVLT)|(^VF)|(^Clock)|(^TM)|(Score)'
df_v1 = df.loc[:, ~df.columns.str.contains(filter_1)]
list(df_v1.columns)

In [ ]:
df_v1.shape

Let's try to filter on variables with many missing values

In [ ]:
missing_values_counts_v1 = df_v1.isna().sum().sort_values(ascending=False)
missing_values_counts_v1.shape #checking

In [ ]:
fig = px.bar(missing_values_counts_v1, height=800, width=2000, labels={'index': 'features'})
fig.update_layout(font_size=10, bargroupgap=0.3)
fig.show()

In [ ]:
len(df_v1)*1/3

In [ ]:
df_v1.shape

In [ ]:
(missing_values_counts_v1 > len(df_v1)*1/3).sum() 
#34 features have a more than 29 missing values

In [ ]:
df_v2 = df_v1.dropna(thresh=(len(df_v1)*2/3), axis='columns')

In [ ]:
missing_values_counts_v2 = df_v2.isna().sum().sort_values(ascending=False)
missing_values_counts_v2.shape #checking

In [ ]:
fig = px.bar(missing_values_counts_v2, height=800, width=2000, labels={'index95': 'features'})
fig.update_layout(font_size=10, bargroupgap=0.3)
fig.show()

In [ ]:
df_desc[df_desc['SUBJECT NUMBER'].isin(df_v2.columns.to_list())] #missing variable description

"MRI YES OR NO" & "Eye Scan (Y/N)" can be excluded as well

In [ ]:
df_v3 = df_v2.drop(axis='columns', labels=["MRI YES OR NO", "Eye Scan (Y/N)"])

In [ ]:
print(df_v3.shape)
print(df_v2.shape)

Because we want to merge some datasets together, only unrelevant features are excluded here. Further cleaning and exploration is performed on merge.ipynb

In [ ]:
df_v3.to_csv("../data/cerebral_perfusion_data/exploration/cerebral_perfusion_diabetes_v1.csv", index=False)